### ISSUE TITLE
Runtime error coercing optional array in zip/scatter

### ISSUE THREAD

**Issue by Ben-Habermeyer** [dnanexus/dxCompiler#358]
Hi DNAnexus,

When running a workflow compiled with dxCompiler 2.10.2, I got the following runtime error at this scatter step (no scatter jobs kicked off)

`value V_Null cannot be coerced to type T_Float at 131:36-133:7 in <string>`
Where `align_initial_bam.outbam` `clip_dd_ds_bam.clipped_bam` are `Array[File]` and `extract_umi.downsample_fraction` is `Array[Float?]`
```
scatter (initial_final_bam_downsample in zip(zip(align_initial_bam.outbam, clip_dd_ds_bam.clipped_bam), extract_umi.downsample_fraction)) {
    call metrics_tasks.sample_qc_metrics {
        input:
            sample=basename(initial_final_bam_downsample.left.left, ".initial.bam"),
            assay_subtype=assay_subtype,
            initial_bam=initial_final_bam_downsample.left.left,
            final_bam=initial_final_bam_downsample.left.right,
            downsample_fraction=initial_final_bam_downsample.right
    }
}
```

Explicitly declaring the result of the zip to a variable seems to fix this issue, so there seems to be an issue with coercing optional array types in zip?
```
    Array[Pair[Pair[File, File], Float?]] initial_final_bam_downsampled = zip(zip(align_initial_bam.outbam, clip_dd_ds_bam.clipped_bam), extract_umi.downsample_fraction)
    scatter (initial_final_bam_downsample in initial_final_bam_downsampled) {
        call metrics_tasks.sample_qc_metrics {
            input:
                sample=basename(initial_final_bam_downsample.left.left, ".initial.bam"),
                assay_subtype=assay_subtype,
                initial_bam=initial_final_bam_downsample.left.left,
                final_bam=initial_final_bam_downsample.left.right,
                downsample_fraction=initial_final_bam_downsample.right
        }
    }
```

Please let me know what you think of this, thank you

**Comment by Gvaihir** [comment]
Hello!
According to WDL spec, [ _all inputs and outputs must be typed_ ](https://github.com/openwdl/wdl/blob/main/versions/1.0/SPEC.md#types). Zip is just another function albeit from the standard library, and its outputs must be typed. Your second version is the right way to declare and type the outputs of your nested `zip`.
Other question is why it's not throwing exceptions at the compile time. This will be in the scope of the respective ticket.

**Comment by Gvaihir** [comment]
I have found examples of undeclared/untyped outputs of the functions in the standard library, but could not find an explicit documentation if this is good practice. Also, most examples were not represented by nested functions. I will check with the language consortium how these cases should be handled.

**Comment by Gvaihir** [comment]
Fixed in [2.10.4](https://github.com/dnanexus/dxCompiler/releases/tag/2.10.4)

**Comment by Ben-Habermeyer** [comment]
nice thank you @Gvaihir !
